### PR TITLE
Add step to run linker on entire runtime pack during libraries build

### PR DESCRIPTION
**Pull request by layomia**
Contributes to https://github.com/dotnet/runtime/issues/38033.

The next [step](https://github.com/dotnet/runtime/pull/40106#discussion_r463049316) following this PR is to suppress individual warnings using baseline xml files. This can be done following https://github.com/mono/linker/pull/1386.

cc @mateoatr @vitek-karas @MichalStrehovsky @sbomer

### CODE REVIEW

File: eng/illink.targets [hunk 1]
```diff
@@ -56,18 +56,18 @@
...
       <ItemName>TrimmingReport</ItemName>
-    </BinPlaceTargetFramework>
-    <BinPlaceTargetFramework Include="$(BuildSettings)">
-      <RuntimePath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/pretrimmed</RuntimePath>
+    </BinPlaceTargetFrameworks>
+    <BinPlaceTargetFrameworks Include="$(NetCoreAppCurrent)">
+      <RuntimePath>$(LibrariesPretrimmedArtifactsPath)</RuntimePath>
       <ItemName>PreTrimmedItem</ItemName>
-    </BinPlaceTargetFramework>
+    </BinPlaceTargetFrameworks>
```
Comment (safern) [inline eng/illink.targets:70]: Remove this one from here, since we will always be binplacing the pre-trimmed assemblies already.

File: src/libraries/Directory.Build.props [hunk 2]
```diff
@@ -328,4 +328,10 @@
     <CLSCompliant Condition="'$(CLSCompliant)' == '' and ('$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true')">false</CLSCompliant>
     <CLSCompliant Condition="'$(CLSCompliant)' == ''">true</CLSCompliant>
   </PropertyGroup>
+
+  <PropertyGroup>
+    <LibrariesPretrimmedArtifactsPath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/pretrimmed</LibrariesPretrimmedArtifactsPath>
+    <LibrariesTrimmedArtifactsPath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/entire_runtimepack_trimmed</LibrariesTrimmedArtifactsPath>
```
Comment (safern) [inline src/libraries/Directory.Build.props:334]:
```suggestion
    <LibrariesTrimmedArtifactsPath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/trimmed-runtimepack</LibrariesTrimmedArtifactsPath>
```

File: src/libraries/Directory.Build.props [hunk 3]
```diff
@@ -328,4 +328,10 @@
     <CLSCompliant Condition="'$(CLSCompliant)' == '' and ('$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true')">false</CLSCompliant>
     <CLSCompliant Condition="'$(CLSCompliant)' == ''">true</CLSCompliant>
   </PropertyGroup>
+
+  <PropertyGroup>
+    <LibrariesPretrimmedArtifactsPath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/pretrimmed</LibrariesPretrimmedArtifactsPath>
```
Comment (safern) [inline src/libraries/Directory.Build.props:333]: Add a common property for `$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)` and use MSBuild::NormalizePath:

```xml
$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'ILLinkTrimAssembly', '$(BuildSettings)'))
```

File: eng/illink.targets [hunk 4]
```diff
@@ -56,18 +56,18 @@
        Must be enabled by setting BinPlaceILLinkTrimAssembly=true
   -->
   <ItemGroup Condition="'$(BinPlaceILLinkTrimAssembly)' == 'true'">
-    <BinPlaceTargetFramework Include="$(BuildSettings)">
+    <BinPlaceTargetFrameworks Include="$(BuildSettings)">
       <RuntimePath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/trimmed</RuntimePath>
```
Comment (safern) [inline eng/illink.targets:60]: Use the common *Root property that I suggested here.

File: eng/illink.targets [hunk 5]
```diff
@@ -56,18 +56,18 @@
...
       <ItemName>TrimmedItem</ItemName>
-    </BinPlaceTargetFramework>
-    <BinPlaceTargetFramework Include="$(BuildSettings)">
+    </BinPlaceTargetFrameworks>
+    <BinPlaceTargetFrameworks Include="$(BuildSettings)">
       <RuntimePath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/reports</RuntimePath>
```
Comment (safern) [inline eng/illink.targets:64]: Ditto.

File: src/libraries/pretest.proj [hunk 6]
```diff
@@ -27,6 +27,19 @@
              Properties="$(TraversalGlobalProperties)" />
   </Target>
 
+  <Target Name="CopyILTrimmedSharedFrameworkLibraries"
```
Comment (safern) [inline src/libraries/pretest.proj:30]: This target is not needed.

File: src/libraries/Directory.Build.targets [hunk 7]
```diff
@@ -64,6 +64,9 @@
     <BinPlaceRef Condition="'$(BinPlaceRef)' == '' and ('$(IsReferenceAssembly)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRef>
     <BinPlaceRuntime Condition="'$(BinPlaceRuntime)' == '' and ('$(IsRuntimeAssembly)' == 'true' or '$(IsRuntimeAndReferenceAssembly)' == 'true')">true</BinPlaceRuntime>
     <BinPlaceForTargetVertical Condition="'$(BinPlaceForTargetVertical)' == ''">true</BinPlaceForTargetVertical>
+
+    <!-- All IsNETCoreApp assemblies will get binplaced to the test sharedframework after the ILLink step. -->
+    <BinPlaceTestSharedFramework Condition="'$(IsNETCoreApp)' == 'true' and '$(BuildAllProjects)' == 'true'">false</BinPlaceTestSharedFramework>
```
Comment (safern) [inline src/libraries/Directory.Build.targets:69]: I think we agreed that we're going to continue binplacing the test shared framework from each individual assembly build, not the whole trimmed shared framework because that would conflict with inner developer experience.

Comment (layomia) [reply src/libraries/Directory.Build.targets:69]: Yes - will remove this change and `CopyILTrimmedSharedFrameworkLibraries` above.

File: src/libraries/illink-sharedframework.proj [hunk 8]
```diff
@@ -0,0 +1,82 @@
+<Project Sdk="Microsoft.Build.Traversal">
```
Comment (ViktorHofer) [inline src/libraries/illink-sharedframework.proj:1]: You are using the Traversal SDK but I don't see any ProjectReferences in this project. Also I don't understand (yet) why we need an extra project for this instead of a targets file.

Comment (ericstj) [reply src/libraries/illink-sharedframework.proj:1]: @ViktorHofer are you recommending to make this a target that is run by the src.proj or some other project instead?  Do you have a recommended project to put this in?

File: eng/illink.targets [hunk 9]
```diff
@@ -52,22 +52,28 @@
...
-    <BinPlaceTargetFramework Include="$(BuildSettings)">
-      <RuntimePath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/pretrimmed</RuntimePath>
+    </BinPlaceTargetFrameworks>
+  </ItemGroup>
+
+  <!-- Binplacing for pre-trimming that is used when running the linker
+       on the entire runtime pack to observe warnings from verbose output
+  -->
+  <ItemGroup>
+    <BinPlaceTargetFrameworks Include="$(NetCoreAppCurrent)">
```
Comment (ericstj) [inline eng/illink.targets:73]: Do we need a separate directory for this?  We can't just use the runtimepack itself?

File: src/libraries/illink-sharedframework.proj (hunk 8 above)
Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.proj:1]: Is there are reason why this can't be a target instead? An extra project which is restored and built as part of src.proj is a bit of an overkill. So yes, a target in src.proj should be sufficient.

I was also commenting on the Traversal SDK being used without any projects being referenced in it which is why I was commenting on this.

File: eng/illink.targets (hunk 9 above)
Comment (safern) [reply eng/illink.targets:73]: I believe the runtime pack contains individually trimmed assemblies. I think they needed non-trimmed assemblies at all.

File: src/libraries/illink-sharedframework.proj (hunk 8 above)
Comment (ericstj) [reply src/libraries/illink-sharedframework.proj:1]: >  a target in src.proj should be sufficient.

Seems reasonable.

> the Traversal SDK being used without any projects being referenced

Probably it wasn't clear which SDK to use that provides basic "build" targets.  In the future, which SDK would you reccomend for this instead?

Comment (safern) [reply src/libraries/illink-sharedframework.proj:1]: I think for those cases we should use the NoTargets SDK.

File: eng/Configurations.props [hunk 10]
```diff
@@ -34,6 +34,7 @@
 
   <!-- Honor the generic RuntimeConfiguration property. -->
   <PropertyGroup>
+    <Configuration Condition="'$(Configuration)' == ''">Debug</Configuration>
```
Comment (eerhardt) [inline eng/Configurations.props:37]: The .NET SDK does this. Why do we need this here?

File: eng/illink.targets [hunk 11]
```diff
@@ -52,22 +52,28 @@
...
+    <BinPlaceTargetFrameworks Include="$(BuildSettings)">
+      <RuntimePath>$([MSBuild]::NormalizePath('$(LibrariesTrimmingArtifactsRoot)', 'reports'))</RuntimePath>
       <ItemName>TrimmingReport</ItemName>
-    </BinPlaceTargetFramework>
-    <BinPlaceTargetFramework Include="$(BuildSettings)">
-      <RuntimePath>$(ArtifactsBinDir)ILLinkTrimAssembly/$(BuildSettings)/pretrimmed</RuntimePath>
+    </BinPlaceTargetFrameworks>
+  </ItemGroup>
+
+  <!-- Binplacing for pre-trimming that is used when running the linker
```
Comment (eerhardt) [inline eng/illink.targets:69]: I'm not sure I understand why we are making these changes. In my mind this shouldn't be changing how we are linking during the library build at all. The library build should stay the same. The only difference is one more step at the end of the build that runs the ILLinker again (on the already trimmed assemblies) but this time with the warnings enabled and the whole shared fx together.

File: src/libraries/illink-sharedframework.proj [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                   'T' from 'System.Lazy<T>' which requires dynamically accessed member kinds 'PublicParameterlessConstructor'
+           IL2009: Could not find method A in type B specified in resource C
+           IL2025: Duplicate preserve of A in B
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
+           IL2035: Unresolved assembly A in DynamicDependencyAttribute on B
+           IL2050: P/invoke method A declares a parameter with COM marshalling. Correctness of COM interop
+                   cannot be guaranteed after trimming. Interfaces and interface members might be removed.
+      -->
+      <ILLinkArgs>$(ILLinkArgs) --nowarn IL2006;IL2009;IL2025;IL2026;IL2035;IL2050</ILLinkArgs>
+    </PropertyGroup>
```
Comment (marek-safar) [inline src/libraries/illink-sharedframework.proj:40]: Is the almost identical copy of these settings intentional?

**Comment by marek-safar** [comment]
Will this obsolete linker run on individual libraries?

**Comment by eerhardt** [comment]
> Will this obsolete linker run on individual libraries?

No, it doesn't. We still run the linker on individual libraries, and that is what ships in the runtimepack. See https://github.com/dotnet/runtime/pull/39133#discussion_r454338996. And the reasoning why we still run the linker on the individual libraries is listed in https://github.com/dotnet/runtime/issues/31712.

File: src/libraries/illink-sharedframework.proj (hunk 8 above)
Comment (layomia) [reply src/libraries/illink-sharedframework.proj:1]: Thanks. I'll change this to a targets file.

File: eng/illink.targets (hunk 9 above)
Comment (layomia) [reply eng/illink.targets:73]: I understand now that we need to run this on the output of the first linker run i.e the runtime pack containing the trimmed assemblies. Will make the change.

Comment (ericstj) [reply eng/illink.targets:73]: Great, glad to reduce the number of binplacing targets.

**Comment by marek-safar** [comment]
> No, it doesn't. We still run the linker on individual libraries, and that is what ships in the runtimepack.

Does it mean we won't get correct warnings for the scenarios which use linker by default?

**Comment by eerhardt** [comment]
> Does it mean we won't get correct warnings for the scenarios which use linker by default?

I'm not sure I follow this question. Can you elaborate?

The scenarios which use linker by default will use the assemblies that are contained in the runtimepack. This change is to pass those same libraries into the linker, now with warnings enabled.

So we are running the linker in 2 different locations during the dotnet/runtime build. Once when building the individual library, like we have always been - warnings disabled. Once at the end of the build against the whole runtimepack - warnings enabled.

File: eng/illink.targets (hunk 11 above)
Comment (layomia) [reply eng/illink.targets]: Yup, thanks for the clarification on running this new step on the trimmed assemblies, not pre-trimmed. The highlighted changes were to create a directory with pre-trimmed assemblies, which we don't need. Will remove.

File: src/libraries/illink-sharedframework.proj (hunk 12 above)
Comment (layomia) [reply src/libraries/illink-sharedframework.proj]: No. I've added a common property group for shared settings.

File: eng/illink.targets [hunk 13]
```diff
@@ -184,23 +184,11 @@
...
-      <ILLinkArgs>$(ILLinkArgs) -c skip</ILLinkArgs>
-      <!-- default action for non-core assemblies -->
-      <ILLinkArgs>$(ILLinkArgs) -u skip</ILLinkArgs>
-      <!-- trim the target assembly -->
-      <ILLinkArgs>$(ILLinkArgs) -p link $(TargetName)</ILLinkArgs>
       <!-- keep type-forward assemblies (facades) -->
       <ILLinkArgs>$(ILLinkArgs) -t</ILLinkArgs>
-      <ILLinkArgs Condition="'$(ILLinkRewritePDBs)' == 'true' and Exists('$(ILLinkTrimAssemblySymbols)')">$(ILLinkArgs) -b true</ILLinkArgs>
       <!-- don't remove the embedded root xml resource since ILLink may run again on the assembly -->
       <ILLinkArgs Condition="'$(ILLinkTrimXml)' != ''">$(ILLinkArgs) --strip-descriptors false</ILLinkArgs>
```
Comment (eerhardt) [inline eng/illink.targets:193]: I don't think these 3 lines are necessary in the "common" set. When linking the whole runtimepack, `$(ILLinkTrimXml)`, `$(ILLinkTrimXmlLibraryBuild)`, and `$(ILLinkSubstitutionsXml)` should never be set, so these 3 lines aren't doing anything in those cases.

File: eng/illink.targets [hunk 14]
```diff
@@ -214,20 +202,41 @@
...
+           IL2009: Could not find method A in type B specified in resource C
+           IL2012: Could not find field A in type B specified in resource C
+           IL2025: Duplicate preserve of A in B
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
+           IL2035: Unresolved assembly A in DynamicDependencyAttribute on B
+           IL2050: P/invoke method A declares a parameter with COM marshalling. Correctness of COM interop
+                   cannot be guaranteed after trimming. Interfaces and interface members might be removed.
       -->
-      <ILLinkArgs>$(ILLinkArgs) --nowarn IL2006;IL2008;IL2009;IL2012;IL2025;IL2026;IL2035;IL2041</ILLinkArgs>
+      <ILLinkArgs>$(ILLinkArgs) --nowarn IL2006;IL2008;IL2009;IL2012;IL2025;IL2026;IL2035;IL2050</ILLinkArgs>
```
Comment (eerhardt) [inline eng/illink.targets:239]: Why the change from `IL2041` to `IL2050`? Nothing in this change should have changed what warnings are produced for the individual library build. So I think both `2041` and `2050` can be removed here.

File: src/libraries/Directory.Build.props [hunk 15]
```diff
@@ -328,4 +328,9 @@
     <CLSCompliant Condition="'$(CLSCompliant)' == '' and ('$(IsTestProject)' == 'true' or '$(IsTestSupportProject)' == 'true')">false</CLSCompliant>
     <CLSCompliant Condition="'$(CLSCompliant)' == ''">true</CLSCompliant>
   </PropertyGroup>
+
+  <PropertyGroup>
+    <LibrariesTrimmedArtifactsPath>$([MSBuild]::NormalizePath('$(ArtifactsBinDir)', 'ILLinkTrimAssembly', '$(BuildSettings)', 'trimmed-runtimepack'))</LibrariesTrimmedArtifactsPath>
```
Comment (eerhardt) [inline src/libraries/Directory.Build.props:333]: Does this property have to go into the common props? It is only used in one place in `illink-sharedframework.targets`. Can it just be moved to `illink-sharedframework.targets`, so it is closer to where it is used, and it doesn't clutter the other projects with unnecessary properties?

**Comment by marek-safar** [comment]
> So we are running the linker in 2 different locations during the dotnet/runtime build. Once when building the individual library, like we have always been - warnings disabled. Once at the end of the build against the whole runtimepack - warnings enabled.

Running linker against the whole runtime pack will produce different output than running linker when building individual libraries? Why are we doing the first run which is now redundant and only slows down the build?

File: src/libraries/illink-sharedframework.targets [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+           IL2025: Duplicate preserve of A in B
+           IL2026: Calling A which has B can break functionality when trimming application code. The target method might be removed.
+           IL2035: Unresolved assembly A in DynamicDependencyAttribute on B
+           IL2050: P/invoke method A declares a parameter with COM marshalling. Correctness of COM interop
+                   cannot be guaranteed after trimming. Interfaces and interface members might be removed.
+      -->
+      <ILLinkArgs>$(ILLinkArgs) --nowarn IL2006;IL2009;IL2025;IL2026;IL2035;IL2050</ILLinkArgs>
+    </PropertyGroup>
+
+     <!-- Retrieve CoreLib's targetpath via GetTargetPath as it isn't binplaced yet. -->
```
Comment (eerhardt) [inline src/libraries/illink-sharedframework.targets:25]: I'm not sure I understand this. Why can't we use `$(MicrosoftNetCoreAppRuntimePackNativeDir)` instead?

Comment (layomia) [reply src/libraries/illink-sharedframework.targets:25]: Because `$(MicrosoftNetCoreAppRuntimePackNativeDir` is populated in the [pretest project](https://github.com/dotnet/runtime/blob/809a06f45161ae686a06b9e9ccc2f45097b91657/src/libraries/pretest.proj#L96) (which occurs after the execution of this step) and it doesn't seem right to add this target (`ILLinkTrimSharedFramework`) to that project.

File: eng/illink.targets (hunk 14 above)
Comment (layomia) [reply eng/illink.targets:239]: I started getting `2050` after syncing with master at the beginning of this PR. Some new code since the initial suppressions check in must have been added which causes these errors:

```
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.CreateBindCtx(UInt32,IBindCtx&): P/invoke method 'System.Runtime.InteropServices.Marshal.CreateBindCtx(UInt32,IBindCtx&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.CreateBindCtx(UInt32,IBindCtx&): P/invoke method 'System.Runtime.InteropServices.Marshal.CreateBindCtx(UInt32,IBindCtx&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.MkParseDisplayName(IBindCtx,String,UInt32&,IMoniker&): P/invoke method 'System.Runtime.InteropServices.Marshal.MkParseDisplayName(IBindCtx,String,UInt32&,IMoniker&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.MkParseDisplayName(IBindCtx,String,UInt32&,IMoniker&): P/invoke method 'System.Runtime.InteropServices.Marshal.MkParseDisplayName(IBindCtx,String,UInt32&,IMoniker&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.BindMoniker(IMoniker,UInt32,Guid&,Object&): P/invoke method 'System.Runtime.InteropServices.Marshal.BindMoniker(IMoniker,UInt32,Guid&,Object&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: System.Runtime.InteropServices.Marshal.BindMoniker(IMoniker,UInt32,Guid&,Object&): P/invoke method 'System.Runtime.InteropServices.Marshal.BindMoniker(IMoniker,UInt32,Guid&,Object&)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\coreclr\src\System.Private.CoreLib\System.Private.CoreLib.csproj]
ILLink : error IL2050: Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VarNumFromParseNum(Byte[],Byte[],Int32): P/invoke method 'Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VarNumFromParseNum(Byte[],Byte[],Int32)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\libraries\Microsoft.VisualBasic.Core\src\Microsoft.VisualBasic.Core.vbproj]
ILLink : error IL2050: Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VarNumFromParseNum(Byte[],Byte[],Int32): P/invoke method 'Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VarNumFromParseNum(Byte[],Byte[],Int32)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\libraries\Microsoft.VisualBasic.Core\src\Microsoft.VisualBasic.Core.vbproj]
ILLink : error IL2050: Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VariantChangeType(Object&,Object&,Int16,Int16): P/invoke method 'Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VariantChangeType(Object&,Object&,Int16,Int16)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\libraries\Microsoft.VisualBasic.Core\src\Microsoft.VisualBasic.Core.vbproj]
ILLink : error IL2050: Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VariantChangeType(Object&,Object&,Int16,Int16): P/invoke method 'Microsoft.VisualBasic.CompilerServices.UnsafeNativeMethods.VariantChangeType(Object&,Object&,Int16,Int16)' declares a parameter with COM marshalling. Correctness of COM interop cannot be guaranteed after trimming. Interfaces and interface members might be removed. [D:\repos\dotnet_runtime_2\src\libraries\Microsoft.VisualBasic.Core\src\Microsoft.VisualBasic.Core.vbproj]
    0 Warning(s)
    10 Error(s)
```

`2041` doesn't show up anymore following the sync hence removed.

**Comment by eerhardt** [comment]
> Running linker against the whole runtime pack will produce different output than running linker when building individual libraries?

Yes, it will produce different warnings. Please read the discussion here https://github.com/dotnet/runtime/pull/39133#discussion_r454338996.

> Why are we doing the first run which is now redundant and only slows down the build?

It is not redundant. Please see read the discussion in #31712 for why we are doing the first run.

File: eng/illink.targets (hunk 14 above)
Comment (eerhardt) [reply eng/illink.targets:239]: Looks like that came in last week with https://github.com/mono/linker/pull/1382.

cc @MichalStrehovsky

File: src/libraries/illink-sharedframework.targets (hunk 16 above)
Comment (eerhardt) [reply src/libraries/illink-sharedframework.targets:29]: My naive thinking is that there should be somewhere in the build (that doesn't rely on tests) that we can hook in where the entire runtimepack is built. But I'm not seeing where that place is. Maybe @safern @ViktorHofer @joperezr or @ericstj might know.

But for now, what you have works so let's go with that.

**Review by eerhardt — APPROVED** [review]
:shipit:

File: src/libraries/illink-sharedframework.targets (hunk 16 above)
Comment (safern) [reply src/libraries/illink-sharedframework.targets:29]: SPC and the runtime gets binplaced into libraries runtime pack and testhost as part of runtime.depproj. 

Runtime.depproj is built as part of pretest.proj which is the last step of the `libs` subset. Which happens after `libs.src` that is the subset that builds `src.proj`. That is why `wasm.targets` also follows that approach.

FWIW we could introduce a common target to get SPC target path instead of duplicating the logic to call MSBuild on it with GetTargetPath target.

Comment (eerhardt) [reply src/libraries/illink-sharedframework.targets:29]: > FWIW we could introduce a common target to get SPC target path instead of duplicating the logic to call MSBuild on it with GetTargetPath target.

It feels to me that the correct fix here would be to fully populate the runtimepack earlier, and not make it dependent on testing.

Comment (safern) [reply src/libraries/illink-sharedframework.targets:29]: The reason why having the runtime binplace at pretest is handy, is for inner loop development. For example if I update code in SPC after a vertical build, in order to run tests against the new SPC, I don't have to do a full libraries vertical, I can just do: `build libs.pretest -rc <runtimeConfig>` and we actually use this in CI to avoid building a vertical against each runtime we run tests (mono, checked coreclr, and release coreclr) and just replace the runtime with the one we want to run tests against by building the pretest subset.

Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.targets:29]: > It feels to me that the correct fix here would be to fully populate the runtimepack earlier, and not make it dependent on testing.

I agree with @eerhardt, we have talked about this couple times already. I believe we should binplace the runtime artifacts incrementally after they are built (CoreLib, coreclr.dll, etc). For CI we just need a target that re-binplaces the runtime artifacts (to preserve the existing behavior which allows to build coreclr and libraries in parallel).

> The reason why having the runtime binplace at pretest is handy, is for inner loop development. For example if I update code in SPC after a vertical build, in order to run tests against the new SPC, I don't have to do a full libraries vertical, I can just do: build libs.pretest -rc <runtimeConfig>

I don't understand why binplacing the runtime artifacts in pretest is handy for inner loop development instead of binplacing the runtime assets at the time they are (re-)built, which doesn't require a libraries rebuild.

**Comment by ViktorHofer** [comment]
Curious, how much time does this additional step take?

File: src/libraries/illink-sharedframework.targets [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+
+    <ILLink AssemblyPaths="$(_AssemblyPaths)"
+        RootAssemblyNames=""
+        OutputDirectory="$(LibrariesTrimmedArtifactsPath)"
+        ExtraArgs="$(ILLinkArgs)"
+        ToolExe="$(_DotNetHostFileName)"
+        ToolPath="$(_DotNetHostDirectory)" />
+  </Target>
+
+  <Import Project="$([MSBuild]::NormalizePath('$(RepositoryEngineeringDir)', 'illink.targets'))" />
```
Comment (ViktorHofer) [inline src/libraries/illink-sharedframework.targets:75]: nit: no need to use NormalizePath here as RepositoryEngineeringDir already returns a trailing slash.

File: src/libraries/illink-sharedframework.targets (hunk 16 above)
Comment (safern) [reply src/libraries/illink-sharedframework.targets:29]: > I don't understand why binplacing the runtime artifacts in pretest is handy for inner loop development instead of binplacing the runtime assets at the time they are (re-)built, which doesn't require a libraries rebuild.

We've also talked about this a couple of times. If you build for example System.Runtime which has a P2P to SPC, and don't specify RuntimeConfiguration as an argument you would potentially be binplacing a `Debug` build of SPC when probably you built the runtime for release, this could cause a mismatch in between the native bits and the managed bits of coreclr which is not recommended:
https://github.com/dotnet/runtime/pull/38885#issuecomment-655082212

So we should be careful on how we do that.

File: src/libraries/src.proj [hunk 18]
```diff
@@ -49,6 +49,9 @@
              Properties="$(TraversalGlobalProperties)" />
   </Target>
 
+  <Import Condition="'$(BuildTargetFramework)' == '$(NetCoreAppCurrent)'"
```
Comment (ViktorHofer) [inline src/libraries/src.proj:52]: nit: "`$(BuildNetCoreAppVertical)` == `true`"

File: src/libraries/illink-sharedframework.targets (hunk 16 above)
Comment (safern) [reply src/libraries/illink-sharedframework.targets:29]: Btw, I agree we should change this if possible, I'm just saying we should be really careful how and where we put this binplacing to happen.

File: src/libraries/illink-sharedframework.targets [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+    </ItemGroup>
+
+    <PropertyGroup>
+      <ILLinkArgs>$(ILLinkArgs) -r @(_AssembliesToLink->'%(Identity)', ' -r ')</ILLinkArgs>
+    </PropertyGroup>
+
+    <!-- When running from Desktop MSBuild, DOTNET_HOST_PATH is not set.
+      In this case, explicitly specify the path to the dotnet host. -->
+    <PropertyGroup Condition=" '$(DOTNET_HOST_PATH)' == '' ">
+      <_DotNetHostDirectory>$(RepoRoot).dotnet</_DotNetHostDirectory>
```
Comment (ViktorHofer) [inline src/libraries/illink-sharedframework.targets:62]: This directory doesn't always exist, ie if the globally installed SDK matches the version specified in global.json. You should instead use `DotNetRoot` (https://github.com/dotnet/arcade/blob/9deb98dc28354641313039da395ce1b976e03c68/src/Microsoft.DotNet.Arcade.Sdk/tools/RepoLayout.props#L30).

File: src/libraries/illink-sharedframework.targets [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+      <_DotNetHostFileName>dotnet</_DotNetHostFileName>
+      <_DotNetHostFileName Condition=" '$(OS)' == 'Windows_NT' ">dotnet.exe</_DotNetHostFileName>
```
Comment (ViktorHofer) [inline src/libraries/illink-sharedframework.targets:64]: nit: Use GetFilePath('$(DotNetTool)') instead.

**Review by ViktorHofer — COMMENTED** [review]
Curious, what are planning to do with the trimmed assemblies in `LibrariesTrimmedArtifactsPath`?

File: src/libraries/illink-sharedframework.targets (hunk 16 above)
Comment (ViktorHofer) [reply src/libraries/illink-sharedframework.targets:29]: > We've also talked about this a couple of times. If you build for example System.Runtime which has a P2P to SPC, and don't specify RuntimeConfiguration as an argument you would potentially be binplacing a Debug build of SPC when probably you built the runtime for release, this could cause a mismatch in between the native bits and the managed bits of coreclr which is not recommended:
#38885 (comment)
Btw, I agree we should change this if possible, I'm just saying we should be really careful how and where we put this binplacing to happen.

I think we all agree that we should change when we binplace the runtime artifacts and I agree that it should be done with the highest diligence. We should do this as part of https://github.com/dotnet/runtime/issues/38034. 

> Also, I wouldn't like removing the build.pretest binplacing the runtime as it is handy for CI as well as I explained above... if we remove that, then provide another hook to binplace the runtime as an individual step.

That's what I meant with: 
> For CI we just need a target that re-binplaces the runtime artifacts (to preserve the existing behavior which allows to build coreclr and libraries in parallel).

Comment (safern) [reply src/libraries/illink-sharedframework.targets:29]: > That's what I meant with:

I somehow missed that part 🤦

> We should do this as part of #38034.

Sounds good.

File: src/libraries/illink-sharedframework.targets (hunk 19 above)
Comment (eerhardt) [reply src/libraries/illink-sharedframework.targets:62]: Whatever we change here, we should make the same change in eng\illink.targets, since they are duplicates:

https://github.com/dotnet/runtime/blob/b552749906402ec7994150bc9dd8970440dbb9c3/eng/illink.targets#L276-L282

**Comment by eerhardt** [comment]
> Curious, what are planning to do with the trimmed assemblies in LibrariesTrimmedArtifactsPath?

For now, we throw them away - they aren't used. We are only running the linker to get the warnings/analysis.

File: eng/illink.targets (hunk 14 above)
Comment (MichalStrehovsky) [reply eng/illink.targets:239]: Yes, we now detect COM usage and mark it as unsafe. Apps that end up referencing these methods after trimming will always warn.

**Comment by layomia** [comment]
@ViktorHofer, thanks for the feedback. I'll address them shortly in a follow up PR.

**Comment by layomia** [comment]
> Curious, how much time does this additional step take?

A local run took 13.366s.

File: src/libraries/src.proj (hunk 18 above)
Comment (layomia) [reply src/libraries/src.proj:52]: The `BuildNetCoreAppVertical` does not appear to be set at this point in the build.
